src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -70,14 +70,16 @@ public virtual IDisposable Start()
             var server = ServerFactory.Start(_serverInstance,
                 async features =>
                 {
-                    var httpContext = contextFactory.CreateHttpContext(features);
-                    httpContext.ApplicationServices = _applicationServices;
-                    var requestIdentifier = GetRequestIdentifier(httpContext);
-
-                    using (logger.BeginScope("Request Id: {RequestId}", requestIdentifier))
+                    using (var httpContext = contextFactory.CreateHttpContext(features))
```

Comment:
I think you have your stack backwards. 'Application' in that kestrel code block refers to the async lamda you're in here (LN71-83), so FireOnCompleted runs after your using block.